src/ffi/logging.rs

```diff
@@ -0,0 +1,236 @@
+    ) -> std::fmt::Result {
+        let meta = event.metadata();
+        write!(
+            writer,
+            "{:?}:{} ",
+            meta.module_path().unwrap_or("<unknown module>"),
+            meta.line().unwrap_or(0),
+        )?;
+
+        ctx.format_fields(writer.by_ref(), event)?;
+
+        writeln!(writer)
+    }
+}
+
+pub struct FfiCallback {
+    callback: Arc<dyn TelioLoggerCb>,
+}
+impl FfiCallback {
+    fn new(logger: Box<dyn TelioLoggerCb>) -> Self {
+        Self {
+            callback: logger.into(),
+        }
+    }
+}
+
+impl<'a> MakeWriter<'a> for FfiCallback {
+    type Writer = FfiCallbackWriter;
+
+    fn make_writer(&self) -> Self::Writer {
+        unreachable!("make_writer_for is used")
+    }
+
+    fn make_writer_for(&self, meta: &tracing::Metadata<'_>) -> Self::Writer {
+        FfiCallbackWriter {
+            cb: self.callback.clone(),
+            level: (*meta.level()).into(),
+        }
+    }
+}
+
+pub struct FfiCallbackWriter {
+    cb: Arc<dyn TelioLoggerCb>,
+    level: TelioLogLevel,
+}
+
+impl io::Write for FfiCallbackWriter {
+    fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
+        // We can be certain that buf fully represents one full event
+        //
+        // See https://github.com/tokio-rs/tracing/blob/tracing-subscriber-0.3.18/tracing-subscriber/src/fmt/fmt_layer.rs#L975
+        //
+        // Additional test added to double protect from future breakages
+
+        let msg = from_utf8(buf)
+            // Trim could be avoided by removing writeln! in formatter
+            // But this makes it more univesal for the future, then we dicide to
+            // switch to other formats
+            .map(|msg| msg.trim().to_string())
+            .unwrap_or_else(|err| format!("failed to decode trace as utf8 bytes: {err:?}"));
+
+        if let Some(filtered_msg) = filter_log_message(msg) {
+            let _ = self.cb.log(self.level, filtered_msg);
+        }
+
+        Ok(buf.len())
+    }
+
+    fn flush(&mut self) -> io::Result<()> {
+        Ok(())
+    }
+}
+
+struct LogStatus {
+    string: String,
+    counter: u32,
+}
+
+lazy_static::lazy_static! {
+    static ref LAST_LOG_STATUS: Mutex<LogStatus> = {
+        Mutex::new(LogStatus{string: String::default(), counter: 0})
+    };
+}
+
+fn filter_log_message(msg: String) -> Option<String> {
+    let mut log_status = match LAST_LOG_STATUS.lock() {
+        Ok(status) => status,
+        Err(_) => {
+            return None;
+        }
+    };
+
+    if !log_status.string.eq(&msg) {
+        log_status.string = msg.clone();
+        log_status.counter = 0;
+        return Some(msg);
+    }
+
+    if log_status.counter > 0 && log_status.counter % 100 == 0 {
+        log_status.counter += 1;
+        return Some(format!("[repeated 100 times!] {}", msg));
+    }
+
+    if log_status.counter < 10 {
+        log_status.counter += 1;
+        return Some(msg);
+    }
+
+    log_status.counter += 1;
+    None
+}
```

Comment:
It seems that some code is just moved (like this function) - I think it makes things a lot simpler, when such moving is kept in separate commits.

Reply:
Yes that's correct, will take note next time.